packages/lodestar/src/chain/genesis/genesis.ts

```diff
@@ -5,104 +5,128 @@
       config,
       config.types.Eth1Data.defaultValue(),
       getTemporaryBlockHeader(config, getEmptyBlock())
     );
     this.depositTree = config.types.DepositDataRootList.tree.defaultValue();
-    this.config = config;
-    this.eth1 = eth1;
-    this.db = db;
-    this.logger = logger;
+    this.depositCache = new Set<number>();
   }
 
   /**
    * Get eth1 deposit events and blocks and apply to this.state until we found genesis.
    */
-  public async waitForGenesis(): Promise<TreeBacked<BeaconState>> {
-    await this.initialize();
-    const eth1DataStream = await this.eth1.getEth1BlockAndDepositEventsSource();
-    const state = await pipe(eth1DataStream, this.processDepositEvents(), this.assembleGenesisState());
-    await this.eth1.endEth1BlockAndDepositEventsSource();
-    return state;
-  }
+  public async waitForGenesis(): Promise<IGenesisResult> {
+    await this.eth1Provider.validateContract();
+
+    // TODO: Load data from data from this.db.depositData, this.db.depositDataRoot
+    // And start from a more recent fromBlock
+    const blockNumberLastestInDb = this.eth1Provider.deployBlock;
+
+    const blockNumberValidatorGenesis = await this.waitForGenesisValidators(blockNumberLastestInDb);
+
+    const depositsAndBlocksStream = getDepositsAndBlockStreamForGenesis(
+      blockNumberValidatorGenesis,
+      this.eth1Provider,
+      this.eth1Params,
+      this.signal
+    );
 
-  private assembleGenesisState(): (source: AsyncIterable<[Deposit[], Eth1Block]>) => Promise<TreeBacked<BeaconState>> {
-    return async (source) => {
-      for await (const [deposits, block] of source) {
-        applyDeposits(this.config, this.state, deposits, this.depositTree);
-        this.logger.verbose(
-          `Found ${this.depositTree.length} deposits and ` + `${this.state.validators.length} validators so far`
-        );
-        if (!block) {
-          continue;
-        }
-        applyTimestamp(this.config, this.state, block.timestamp);
-        applyEth1BlockHash(this.config, this.state, fromHexString(block.hash));
-        const isValid = isValidGenesisState(this.config, this.state);
-        if (isValid) {
-          this.logger.info(`Found genesis state at eth1 block ${block.number}`);
-          return this.state;
-        }
+    for await (const [depositEvents, block] of depositsAndBlocksStream) {
+      this.applyDeposits(depositEvents);
+      applyTimestamp(this.config, this.state, block.timestamp);
+      applyEth1BlockHash(this.config, this.state, fromHexString(block.hash));
+      if (isValidGenesisState(this.config, this.state)) {
+        this.logger.info(`Found genesis state at eth1 block ${block.number}`);
+        return {
+          state: this.state,
+          depositTree: this.depositTree,
+          block,
+        };
       }
-      throw Error("Did not found genesis state and block source stopped");
-    };
+    }
+
+    throw Error("depositsStream stopped without a valid genesis state");
```

Comment:
I'm pretty sure this error would end up as unhandled rejection, as we don't handle error all the way till chain.start

Reply:
What would you do here instead?

Reply:
I think that's fine, we should probably catch error in chain.start